src/sql/workbench/api/node/extHostNotebook.ts

```diff
@@ -70,14 +73,132 @@ export class ExtHostNotebook implements ExtHostNotebookShape {
 		return this._withContentManager(managerHandle, (contentManager) => contentManager.save(URI.revive(notebookUri), notebook));
 	}
 
+	$refreshSpecs(managerHandle: number): Thenable<sqlops.nb.IAllKernels> {
+		return this._withSessionManager(managerHandle, async (sessionManager) => {
+			await sessionManager.ready;
+			return sessionManager.specs;
+		});
+	}
+
+	$startNewSession(managerHandle: number, options: sqlops.nb.ISessionOptions): Thenable<INotebookSessionDetails> {
+		return this._withSessionManager(managerHandle, async (sessionManager) => {
+			let session = await sessionManager.startNew(options);
+			let sessionId = this._addNewAdapter(session);
+			let kernelDetails: INotebookKernelDetails = undefined;
+			if (session.kernel) {
+				kernelDetails = this.saveKernel(session.kernel);
+			}
+			let details: INotebookSessionDetails = {
+				sessionId: sessionId,
+				id: session.id,
+				path: session.path,
+				name: session.name,
+				type: session.type,
+				status: session.status,
+				canChangeKernels: session.canChangeKernels,
+				kernelDetails: kernelDetails
+			};
+			return details;
+		});
+	}
+
+	private saveKernel(kernel: sqlops.nb.IKernel): INotebookKernelDetails {
+		let kernelId = this._addNewAdapter(kernel);
+		let kernelDetails: INotebookKernelDetails = {
+			kernelId: kernelId,
+			id: kernel.id,
+			info: kernel.info,
+			name: kernel.name,
+			supportsIntellisense: kernel.supportsIntellisense
+		};
+		return kernelDetails;
+	}
+
+	$shutdownSession(managerHandle: number, sessionId: string): Thenable<void> {
+		return this._withSessionManager(managerHandle, async (sessionManager) => {
+			return sessionManager.shutdown(sessionId);
+		});
+	}
+
+	$changeKernel(sessionId: number, kernelInfo: sqlops.nb.IKernelSpec): Thenable<INotebookKernelDetails> {
+		let session = this._getAdapter<sqlops.nb.ISession>(sessionId);
+		return session.changeKernel(kernelInfo).then(kernel => this.saveKernel(kernel));
+	}
+
+	$getKernelReadyStatus(kernelId: number): Thenable<sqlops.nb.IInfoReply> {
+		let kernel = this._getAdapter<sqlops.nb.IKernel>(kernelId);
+		return kernel.ready.then(success => kernel.info);
+	}
+
+	$getKernelSpec(kernelId: number): Thenable<sqlops.nb.IKernelSpec> {
+		let kernel = this._getAdapter<sqlops.nb.IKernel>(kernelId);
+		return kernel.getSpec();
+	}
+
+	$requestComplete(kernelId: number, content: sqlops.nb.ICompleteRequest): Thenable<sqlops.nb.ICompleteReplyMsg> {
+		let kernel = this._getAdapter<sqlops.nb.IKernel>(kernelId);
+		return kernel.requestComplete(content);
+	}
+
+	$requestExecute(kernelId: number, content: sqlops.nb.IExecuteRequest, disposeOnDone?: boolean): Thenable<INotebookFutureDetails> {
+		let kernel = this._getAdapter<sqlops.nb.IKernel>(kernelId);
+		let future = kernel.requestExecute(content, disposeOnDone);
+		let futureId = this._addNewAdapter(future);
+		this.hookFutureDone(futureId, future);
```

Comment:
We need to listen for all messages and send to main thread, which then forwards to any handlers that are registered there.